core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3303,6 +3304,28 @@ class KafkaApis(val requestChannel: RequestChannel,
       new DescribeTransactionsResponse(response.setThrottleTimeMs(requestThrottleMs)))
   }
 
+  def handleListTransactionsRequest(request: RequestChannel.Request): Unit = {
+    val listTransactionsRequest = request.body[ListTransactionsRequest]
+    val response = new ListTransactionsResponseData()
+
+    val filteredProducerIds = listTransactionsRequest.data.producerIdFilter.asScala.map(Long.unbox).toSet
+    val filteredStates = listTransactionsRequest.data.statesFilter.asScala.toSet
+
+    txnCoordinator.handleListTransactions(filteredProducerIds, filteredStates) match {
+      case Left(error) =>
+        response.setErrorCode(error.code)
+      case Right(transactions) =>
+        val authorizedTransactions = transactions.filter { state =>
+          authHelper.authorize(request.context, DESCRIBE, TRANSACTIONAL_ID, state.transactionalId)
```

Comment:
Why `TransactionState` does not include error field (similar to `DescribeTransactionsResponseData.TransactionState`)?

Reply:
In the case of `DescribeTransactions`, the set of transaction states to include in the response are explicitly defined in the request. With `ListTransactions`, we only have filters. The only error state to convey is at the top level. For individual transaction states, they either match the filters or they do not.